bentoml/_internal/server/grpc/servicer.py

```diff
@@ -0,0 +1,189 @@
+        self.mount_servicers = [] if not mount_servicers else list(mount_servicers)
+        self.interceptors = [] if not interceptors else list(interceptors)
+        self.loaded = False
+
+    def load(self):
+        assert not self.loaded
+
+        self.interceptors_stack = self.build_interceptors_stack()
+
+        self.bento_servicer = create_bento_servicer(self.bento_service)
+
+        # Create a health check servicer. We use the non-blocking implementation
+        # to avoid thread starvation.
+        self.health_servicer = health.aio.HealthServicer()
+
+        self.service_names = tuple(
+            service.full_name for service in pb.DESCRIPTOR.services_by_name.values()
+        ) + (health.SERVICE_NAME,)
+        self.loaded = True
+
+    def build_interceptors_stack(self) -> list[aio.ServerInterceptor]:
+        return list(map(lambda x: x(), self.interceptors))
+
+    async def startup(self):
+        for handler in self.on_startup:
+            if is_async_iterable(handler):
+                await handler()
+            else:
+                handler()
+
+    async def shutdown(self):
+        for handler in self.on_shutdown:
+            if is_async_iterable(handler):
+                await handler()
+            else:
+                handler()
+
+    def __bool__(self):
+        return self.loaded
+
+
+def is_async_iterable(obj: t.Any) -> bool:  # pragma: no cover
+    return asyncio.iscoroutinefunction(obj) or (
+        callable(obj) and asyncio.iscoroutinefunction(obj.__call__)
+    )
+
+
+def create_bento_servicer(service: Service) -> services.BentoServiceServicer:
+    """
+    This is the actual implementation of BentoServicer.
+    Main inference entrypoint will be invoked via /bentoml.grpc.<version>.BentoService/Call
+    """
+    from ...io_descriptors.multipart import Multipart
+
+    class BentoServiceImpl(services.BentoServiceServicer):
+        """An asyncio implementation of BentoService servicer."""
+
+        async def Call(  # type: ignore (no async types) # pylint: disable=invalid-overridden-method
+            self,
+            request: pb.Request,
+            context: BentoServicerContext,
+        ) -> pb.Response | None:
+            if request.api_name not in service.apis:
+                raise UnprocessableEntity(
+                    f"given 'api_name' is not defined in {service.name}",
+                ) from None
+
+            api = service.apis[request.api_name]
+            response = pb.Response()
+
+            try:
+                _use_internal_bytes_contents = False
+                # map by default is optional ,and hence we need to check
+                # whether multipart is an empty dict or not.
+                if len(request.multipart) != 0:
+                    io_proto = request.multipart
+                elif request.HasField(api.input.accepted_proto_fields):
+                    io_proto = getattr(request, api.input.accepted_proto_fields)
```

Comment:
Wait, is `accepted_proto_fields` not a list? How does this work if an IO descriptor accepts multiple fields?

Reply:
i forgot to change this to a list

Reply:
let me do that

Reply:
actually no thinking about this, this would probably make it a bit more complicated. I will make an issue for this and address this later